src/model.rs

```diff
@@ -0,0 +1,768 @@
+use yew::events::{IKeyboardEvent, ClickEvent, KeyPressEvent};
+use yew::services::{ConsoleService};
+use yew::services::reader::{File, FileData, ReaderService, ReaderTask};
+use yew::virtual_dom::{VList};
+use pest::Parser;
+use std::fs;
+use std::panic;
+
+use crate::grammar::{Grammar, Kind, Interactive};
+use crate::style::Style;
+use crate::coordinate::{Coordinate, Row, Col};
+use crate::session::Session;
+use crate::utils::{
+    resize_cells, 
+    resize, 
+    apply_definition_grammar, 
+    non_zero_u32_tuple, 
+    move_grammar
+};
+use crate::{
+    row_col_vec, 
+    coord, 
+    coord_row, 
+    coord_col
+};
+
+
+#[derive(Parser)]
+#[grammar = "coordinate.pest"]
+pub struct CoordinateParser;
+
+
+// Model contains the entire state of the application
+#[derive(Debug)]
+pub struct Model {
+    // model holds a direct reference to the topmost root A1 and meta A2 grammars
+    // these two grammars are excluded from the grammar Map
+    root: Grammar,
+    meta: Grammar,
+
+    // the view that the UI treats as the topmost grammar to start rendering from.
+    view_root: Coordinate,
+
+    pub grammars: HashMap</*Key*/ Coordinate, /*Value*/ Grammar>,
+    value: String,  // what is this?????
+    active_cell: Option<Coordinate>,
+    suggestions: Vec<Coordinate>,
+
+    pub col_widths: HashMap<Col, f64>,
+    pub row_heights: HashMap<Row, f64>,
+
+    // tabs correspond to sessions
+    tabs: Vec<String>,
+    current_tab: i32,
+
+    // side menus
+    side_menus: Vec<SideMenu>,
+    open_side_menu: Option<i32>,
+
+    console: ConsoleService,
+    reader: ReaderService,
+
+    link: ComponentLink<Model>,
+    tasks: Vec<ReaderTask>,
+}
+
+#[derive(Debug)]
+struct SideMenu {
+    name: String,
+    icon_path: String,
+}
+
+// ACTIONS
+// Triggered in the view, sent to update function
+pub enum Action {
+    // Do nothing
+    Noop,
+
+    // Change string value of Input grammar
+    ChangeInput(Coordinate, /* new_value: */ String),
+
+    // Show suggestions dropdown at Coordinate based on query
+    ShowSuggestions(Coordinate, /* query: */ String),
+
+    SetActiveCell(Coordinate),
+
+    DoCompletion(/* source: */ Coordinate, /* destination */ Coordinate),
+
+    SetActiveMenu(Option<i32>),
+
+    ReadSession(/* filename: */ File),
+
+    LoadSession(FileData),
+
+    SaveSession(),
+
+    // Grid Operations
+    AddNestedGrid(Coordinate, (u32 /*rows*/, u32 /*cols*/)),
+
+    InsertRow,
+    InsertCol,
+
+    // Alerts and stuff
+    Alert(String),
+}
+
+impl Model {
+    fn load_session(&mut self, session: Session) {
+        self.root = session.root;
+        self.meta = session.meta;
+        self.grammars = session.grammars;
+    }
+
+    fn to_session(&self) -> Session {
+        Session {
+            root: self.root.clone(),
+            meta: self.meta.clone(),
+            grammars: self.grammars.clone(),
+        }
+    }
+
+    fn get_style(&self, coord: &Coordinate) -> String {
```

Comment:
remove this from the model impl and add it to style.rs as a stand-alone function. don’t forget to replace references to “self” with the concrete “Model” type

Reply:
will do.